ykrt/src/compile/jitc_yk/trace_builder.rs

```diff
@@ -615,12 +642,115 @@ impl<'a> TraceBuilder<'a> {
         let instr = match cast_kind {
             aot_ir::CastKind::SignExtend => jit_ir::SignExtendInst::new(
                 &self.handle_operand(val)?,
-                self.handle_type(*dest_type_idx)?,
+                self.handle_type(self.aot_mod.type_(*dest_type_idx))?,
             ),
         };
         self.copy_instruction(instr.into(), bid, aot_inst_idx)
     }
 
+    fn handle_switch(
+        &mut self,
+        bid: &aot_ir::BBlockId,
+        aot_inst_idx: usize,
+        sm: &'a aot_ir::Instruction,
+        next_bb: &aot_ir::BBlockId,
+        test_val: &aot_ir::Operand,
+        _default_dest: &aot_ir::BBlockIdx,
+        case_values: &Vec<u64>,
+        case_dests: &Vec<aot_ir::BBlockIdx>,
+    ) -> Result<(), CompilationError> {
+        if case_values.len() == 0 {
+            // Degenerate switch. Not sure it can even happen.
+            panic!();
+        }
+
+        // Find the JIT type of the value being tested.
+        let jit_type_idx = self.handle_type(test_val.type_(self.aot_mod))?;
+        let jit_int_type = match self.jit_mod.type_(jit_type_idx) {
+            jit_ir::Ty::Integer(it) => it,
+            _ => unreachable!(),
+        }
+        .to_owned();
+
+        // Find out which case we traced.
+        let guard = match case_dests.iter().position(|&cd| cd == next_bb.bb_idx()) {
+            Some(cidx) => {
+                // A non-default case was traced.
+                let val = case_values[cidx];
+                let bb = case_dests[cidx];
+
+                // Build the constant value to guard.
+                let jit_const = jit_int_type
+                    .to_owned()
+                    .make_constant(&mut self.jit_mod, val)?;
+                let jit_const_opnd = jit_ir::Operand::Const(self.jit_mod.const_idx(&jit_const)?);
+
+                // Perform the comparison.
+                let jit_test_val = self.handle_operand(test_val)?;
+                let cmp_inst =
+                    jit_ir::IcmpInst::new(jit_test_val, jit_ir::Predicate::Equal, jit_const_opnd);
+                let jit_cond = self.jit_mod.push_and_make_operand(cmp_inst.into())?;
+
+                // Guard the result of the comparison.
+                self.create_guard(&jit_cond, bb == next_bb.bb_idx(), sm)?
+            }
+            None => {
+                // The default case was traced.
+                //
+                // We need a guard that expresses that `test_val` was not any of the `case_values`.
+                // We encode this in the JIT IR like:
+                //
+                //     member = case_val1 || ... || case_valN
+                //     guard member == 0
+                //
+                // OPT: This could be much more efficient if we introduced a special "none of these
```

Comment:
I don't think a "none of these values" values is a good idea. It's really tricky to do this correctly and it will often bloat memory.

Reply:
That's the semantics we have to implement I'm afraid.

If you have:
```
switch x {
case 0:
case 147:
case 5:
case 8888:
    break;
default:
    blah...
}
```

And we trace the default branch, then we must guard that `x` wasn't 0, 147, 5, of 8888.

Reply:
Note that this PR already (inefficiently) implements these semantics.

Reply:
?

Reply:
My point is we can't make a "value that is none of these values". Nowt to do with semantics :)

Reply:
I didn't understand this, but I think we resolved this offline?

Reply:
Ah! Gotcha!